### ISSUE TITLE
Possible typedef regression in 0.23

### ISSUE THREAD

**Issue by OtaK** [mozilla/uniffi-rs#1466]
Hi!

I tried updating to 0.23 (On this commit: 2d8c65856746952ce9004b3bd6ccce6914ef7f71 as I need the `bindgen` feature) and I'm currently running into the following issue; 

### Context 

We currently have several Rust typedefs like these: 

```rust
pub type JsonRawData = Vec<u8>;
```

A lot of our internal data are just opaque buffers to our library so we try to have more semantic/contextual names even if it ends up resolving to a `Vec<u8>` anyway.

Then we expose those over UniFFI like so: 

```
[Custom]
typedef sequence<u8> JsonRawData;
```

### Problems

Before 0.23, this was compiling fine and was exposed correctly.
After 0.23, I get two different compilation errors.

Keeping the `[Custom]` tag creates the following error

```
error[E0119]: conflicting implementations of trait `FfiConverter<UniFfiTag>` for type `Vec<u8>`
    --> /[workspace]/out/MyLibrary.uniffi.rs:3204:1
     |
3204 | unsafe impl uniffi::FfiConverter<crate::UniFfiTag> for r#JsonRawData {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: conflicting implementation in crate `uniffi_core`:
             - impl<UT, T> FfiConverter<UT> for Vec<T>
               where T: FfiConverter<UT>;
```

I guessed the `[Custom]` directive was the issue, because *technically* it indeed isn't following the newtype pattern, so I removed it 

```
typedef sequence<u8> JsonRawData;
```

And now I get: 

```
error: failed to run custom build command for `core-crypto-ffi v0.6.0-rc.5 (/home/otak/Dev/wire/corelib/crypto-ffi)`

Caused by:
  process didn't exit successfully: `/[workspace]/ffi-29b24dd8df415572/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=./src/MyLibrary.udl
  cargo:rerun-if-env-changed=UNIFFI_TESTS_DISABLE_EXTENSIONS

  --- stderr
  thread 'main' panicked at 'must have a crate name', /home/otak/.cargo/git/checkouts/uniffi-rs-6f89edd2a1ffa4bd/2d8c658/uniffi_bindgen/src/interface/attributes.rs:425:14
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Digging into the code, this seems to indicate that the parser automatically assumes that `[External]` is present and thus requires the form `[External="crate_name"]`, while it is not. 

I also tried to trick `uniffi` by doing `[External="crate"]` (and other variations with `super`, `my_library` or an empty string) but nothing works.

**Comment by mhammond** [comment]
I think this was caused by reorganizing how the FfiConverters are managed. I suspect that if you had more than 1 type alias pre 0.23 (eg, `pub type JsonRawData = Vec<u8>;` *and* `pub type BufferRawData = Vec<u8>;` in the same crate) you would have struck it there too?

What *does* appear to still work fine is something like `struct JsonRawData(Vec<u8>)`, which was really the intent behind custom types (ie, that they are a real types rather than an alias) - IOW, I think the fact that worked before was accidental rather than intentional. Hopefully others will chime in here though.

**Comment by mhammond** [comment]
This might also be an argument for a "real" typedef though - ie, I can't see why it wouldn't be possible to support that typedef without any attributes and have uniffi also just treat it as an alias - it's just that we don't support that currently as [noted here](https://github.com/mozilla/uniffi-rs/blob/main/uniffi_bindgen/src/interface/types/finder.rs#L95)

**Comment by OtaK** [comment]
>  I suspect that if you had more than 1 type alias pre 0.23 (eg, pub type JsonRawData = Vec<u8>; and pub type BufferRawData = Vec<u8>; in the same crate) you would have struck it there too?

No, we have exactly 5 typedefs that resolve to `pub type MyType = Vec<u8>;` and we never had a problem with those; IIRC we've been using UniFFI ever since version 0.15 or 0.16 and it always worked.

https://github.com/wireapp/core-crypto/blob/develop/crypto-ffi/src/CoreCrypto.udl#L293-L300

I thought about using simple wrapped newtypes (with something like `#[repr(transparent)]`, this will work for sure!

If there was never any intent for this usecase to work - I think we should document that type aliases are *not* supported and that instead a solution for it is a simple wrapped newtype.
